### DISCUSSION TITLE
Is it possible to make video calls using Wi-Fi Direct in a p2p connection?

### DISCUSSION THREAD

**Discussion by SatoshiKamiya** [flutter-webrtc/flutter-webrtc#1665] [Q&A]
Nice to meet you. I would like to implement video calls between tablets using the flutter-webrtc plugin.

The setup is as follows: We use a router with an active internet connection as an access point, and both tablet devices are connected to this access point via WiFi. Under these conditions, I was able to confirm that video calling via WebRTC worked.

However, without the router, in an environment where both devices have no internet connection and are connected only via WiFi Direct, I was unable to make video calls.

Offer and answer messages were generated, and I confirmed that the devices were able to send and receive messages between each other. Both devices were able to obtain their own IP addresses in the WiFi Direct environment. However, only local IP addresses were displayed as connection candidates.

Is it impossible to make video calls using WebRTC in a P2P connection via WiFi Direct?

**Comment by cloudwebrtc** [comment]
I think we can use packages like nsd to discover devices in the LAN, and then use a simple C/S mode to establish a socket to exchange offer/answer to make the webrtc connection establishment.

https://pub.dev/packages/nsd

**Reply by SatoshiKamiya** [reply]
Thank you for your prompt reply. I'll give it a try.

**Reply by SatoshiKamiya** [reply]
Thank you for your continued support.
 I followed the advice I received, but it didn’t work.
  The steps I took are as follows:

Connected the tablets via WiFi-Direct Registered NSD on each tablet Confirmed the registration via logs
Attempted a WebRTC video call (no candidates appeared in the connection process...)

I used the Flutter plugin "nsd:3.0.0" for NSD registration. The registration process is as follows:

code:
final registration = await register(const Service(name: 'test', type: '_http._tcp', port: 8080));
print("[NSD check] name = ${registration.service.name}");
print("[NSD check] type = ${registration.service.type}");
print("[NSD check] port = ${registration.service.port}");

Log:
2024-09-11 17:52:13.509 16795-16841 flutter                 com.example.base                     I  [NSD check] name = test
2024-09-11 17:52:13.509 16795-16841 flutter                 com.example.base                     I  [NSD check] type = _http._tcp
2024-09-11 17:52:13.509 16795-16841 flutter                 com.example.base                     I  [NSD check] port = 8080

Is it still impossible to establish a connection even when using NSD?
 Or is there a mistake in the way I’m implementing it?

**Reply by cloudwebrtc** [reply]
I think NSD just announces that you have a service on port 8080. I guess you actually need to write a dart http server or websocket server to listen on port 8080 so that another client that discovers this service can establish a real http/ws connection.

**Reply by SatoshiKamiya** [reply]
Thank you again for your continuous advice.

As you pointed out, I was connected to both Wi-Fi Direct and WebSocket at the same time.
Afterward, I had registered with SND, so after connecting to Wi-Fi Direct, 
I will try to register the SND service and then establish a WebSocket connection based on the registered service.

**Reply by SatoshiKamiya** [reply]
Thank you for the advice you provided the other day.
Following your suggestion, I registered with NSD and then connected via WebSocket. 
After that, I attempted communication using WebRTC, but the connection failed.

After connecting via WiFi-Direct, I checked whether I could obtain the IP address using network_info_plus, 
but the result was null. (When connected to a regular WiFi network, it returns a value.)

I suspect that the information defined by NSD is set on the WiFi-Direct side, and WebRTC is unable to access it. 
What do you think? Do you have any ideas on how to resolve this?

**Comment by SatoshiKamiya** [comment]
I was able to resolve the issue on my own, so I’m reporting it here. 
The Android library in the Flutter plugin was loading implementation 'io.github.webrtc-sdk:android:125.6422.03', and within this JAR file,
I found the NetworkMonitorAutoDetect.class. By writing the setIncludeWifiDirect method from this class in the Android native code, 
it worked. For example, you can write setIncludeWifiDirect(true); somewhere in /flutter_webrtc/android/src/main/java/com/cloudwebrtc/webrtc/MethodCallHandlerImpl.java.
